src/us/kbase/workspace/test/controllers/sample/SampleServiceController.java

```diff
@@ -0,0 +1,229 @@
+			this.nodeEdgeCollection = checkString(nodeEdgeCollection, "nodeEdgeCollection");
+			this.dataLinkCollection = checkString(dataLinkCollection, "dataLinkCollection");
+			this.workspaceObjectVersionShadowCollection = checkString(
+					workspaceObjectVersionShadowCollection,
+					"workspaceObjectVersionShadowCollection");
+			this.schemaCollection = checkString(schemaCollection, "schemaCollection");
+		}
+	}
+
+	/** Create the sample service controller.
+	 * @param arango the ArangoDB controller where the sample service will store files.
+	 * @param sampleServiceDir the directory containing the sample service repo.
+	 * @param rootTempDir a temporary directory in which to store sample service data and log
+	 *  files. The files will be stored inside a child directory that is unique per invocation.
+	 * @param authURL the URL of the KBase authentication service.
+	 * @param authToken a valid authentication service token.
+	 * @param workspaceURL the URL of the KBase workspace service.
+	 * @param workspaceToken an authentication token with admin read privileges to the
+	 *  workspace service.
+	 * @param sampleFullAdminRole an auth role that will grant full admin priviledges to the
+	 *  sample service.
+	 * @param arangoParams parameters for sample service communications with ArangoDB.
+	 * @throws Exception if an error occurs.
+	 */
+	public SampleServiceController(
+			final ArangoController arango,
+			final Path sampleServiceDir,
+			final Path rootTempDir,
+			final URL authURL,
+			final String authToken,
+			final String sampleFullAdminRole,
+			final URL workspaceURL,
+			final String workspaceToken,
+			final SampleServiceArangoParameters arangoParams)
+			throws Exception {
+
+		tempDir = makeTempDirs(rootTempDir, "SampleServiceController-", new LinkedList<String>());
+
+		port = findFreePort();
+		File ssIniFile = createSampleServiceDeployCfg(
+				arango,
+				authURL,
+				authToken,
+				workspaceURL,
+				workspaceToken,
+				sampleFullAdminRole,
+				arangoParams);
+
+		String libDir = "lib";
+		Path libRoot = tempDir.resolve(libDir);
+		FileUtils.copyDirectory(sampleServiceDir.toFile(), libRoot.toFile());
+
+		final String libDirPath = libRoot.toAbsolutePath().toString();
+		logfile = tempDir.resolve("sample_service.log");
+		ProcessBuilder samplepb = new ProcessBuilder(
+				"gunicorn",
+				"--worker-class", "gevent",
+				"--timeout", "30",
+				"--workers", "1",
+				"--bind", ":" + port,
+				"--log-level", "info",
+				String.format("%s.%sServer:application", SAMPLE_SERVICE_NAME, SAMPLE_SERVICE_NAME))
+				.redirectErrorStream(true)
+				.redirectOutput(logfile.toFile());
+		Map<String, String> env = samplepb.environment();
+		env.put("KB_DEPLOYMENT_CONFIG", ssIniFile.getAbsolutePath().toString());
+		env.put("KB_SERVICE_NAME", SAMPLE_SERVICE_NAME);
+		env.put("PYTHONPATH", libDirPath);
+		samplepb.directory(new File(libDirPath));
+		sampleService = samplepb.start();
+
+		Thread.sleep(1000); //let the service start up
```

Comment:
Will there be a problem if it doesn't start up in this time? Anything that should be trapped?

Reply:
The tests will fail with the workspace complaining it can't contact the sample service. I've never seen this happen, 1s appears to be more than enough time.

> Anything that should be trapped?

I don't think so. If this wasn't a test rig my opinion would be different, of course.

Reply:
👍